internal/contour/handler.go

```diff
@@ -55,9 +58,13 @@ type EventHandler struct {
 	// seq is the sequence counter of the number of times
 	// an event has been received.
 	seq int
+
+	syncTracker *synctrack.SingleFileTracker
+
+	initialDagBuilt bool
```

Comment:
You're right; since the value will only be written once, using an atomic boolean might be unnecessary.

Reply:
We should change this, the go race detector picks this up in e2e tests when the contour binary is compiled with the `-race` flag:

```
  time="2023-10-05T18:44:14Z" level=info msg="waiting for the initial dag to be built" context=xds
  time="2023-10-05T18:44:14Z" level=info msg="performing delayed update" context=contourEventHandler last_update=100.922283ms outstanding=1
  time="2023-10-05T18:44:14Z" level=debug msg="cluster load assignments did not change" context=endpointstranslator
  ==================
  WARNING: DATA RACE
  Write at 0x00c0003186f8 by goroutine 354:
    github.com/projectcontour/contour/internal/contour.(*EventHandler).Start()
        /home/sunjayb/workspace/contour/internal/contour/handler.go:208 +0x558
    sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func1()
        /home/sunjayb/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:223 +0x1c1
    sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func2()
        /home/sunjayb/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:226 +0x47

  Previous read at 0x00c0003186f8 by goroutine 362:
    github.com/projectcontour/contour/internal/contour.(*EventHandler).HasBuiltInitialDag()
        /home/sunjayb/workspace/contour/internal/contour/handler.go:115 +0x44
    github.com/projectcontour/contour/internal/contour.(*EventHandler).HasBuiltInitialDag-fm()
        <autogenerated>:1 +0x1d
    main.(*xdsServer).Start.func1()
        /home/sunjayb/workspace/contour/cmd/contour/serve.go:866 +0x47
    k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext.func2()
        /home/sunjayb/go/pkg/mod/k8s.io/apimachinery@v0.28.2/pkg/util/wait/loop.go:73 +0x7b
    k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext()
        /home/sunjayb/go/pkg/mod/k8s.io/apimachinery@v0.28.2/pkg/util/wait/loop.go:74 +0x307
    k8s.io/apimachinery/pkg/util/wait.PollUntilContextCancel()
        /home/sunjayb/go/pkg/mod/k8s.io/apimachinery@v0.28.2/pkg/util/wait/poll.go:33 +0x70
    main.(*xdsServer).Start()
        /home/sunjayb/workspace/contour/cmd/contour/serve.go:865 +0x118
    sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func1()
        /home/sunjayb/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:223 +0x1c1
    sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func2()
        /home/sunjayb/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:226 +0x47

  Goroutine 354 (running) created at:
    sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile()
        /home/sunjayb/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:207 +0x24e
    sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).Start.func1.2()
        /home/sunjayb/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:136 +0x39

  Goroutine 362 (running) created at:
    sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile()
        /home/sunjayb/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:207 +0x24e
    sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).Start.func1.2()
        /home/sunjayb/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:136 +0x39
  ==================
  time="2023-10-05T18:44:14Z" level=info msg="the initial dag is built" context=xds
  time="2023-10-05T18:44:14Z" level=info msg="started xDS server type: \"contour\"" address="0.0.0.0:8001" context=xds insecure=true
```

Reply:
Should be able to use a channel here instead of polling to address this

Reply:
Dear Sunjay, do you run a different command than `make check-test-race`?
I don't see the error when using this target. Could you please provide your command so that I can verify my changes are indeed effective?

Reply:
+ What's your suggested change? I think, as @tsaarni pointed out, using `atomic.Bool` instead of `bool` would do the thing.

Reply:
> I think, as @tsaarni pointed out, using atomic.Bool instead of bool would do the thing.

:+1: that's fine with me

Reply:
here's the makefile diff:
```diff
diff --git a/Makefile b/Makefile
index 385325be..e890d37c 100644
--- a/Makefile
+++ b/Makefile
@@ -324,7 +324,7 @@ e2e: | setup-kind-cluster load-contour-image-kind run-e2e cleanup-kind ## Run E2
 run-e2e:
        CONTOUR_E2E_LOCAL_HOST=$(CONTOUR_E2E_LOCAL_HOST) \
        CONTOUR_E2E_IMAGE=$(CONTOUR_E2E_IMAGE) \
-       go run github.com/onsi/ginkgo/v2/ginkgo -tags=e2e -mod=readonly -skip-package=upgrade,bench -keep-going -randomize-suites -randomize-all -poll-progress-after=120s --focus '$(CONTOUR_E2E_TEST_FOCUS)' -r $(CONTOUR_E2E_PACKAGE_FOCUS)
+       go run github.com/onsi/ginkgo/v2/ginkgo -v -tags=e2e -mod=readonly -skip-package=upgrade,bench -keep-going -randomize-suites -randomize-all -poll-progress-after=120s --focus '$(CONTOUR_E2E_TEST_FOCUS)' -r $(CONTOUR_E2E_PACKAGE_FOCUS)

 .PHONY: cleanup-kind
 cleanup-kind:
```

and the e2e test diff:

```diff
diff --git a/test/e2e/framework.go b/test/e2e/framework.go
index 7ff7e351..4cfd0feb 100644
--- a/test/e2e/framework.go
+++ b/test/e2e/framework.go
@@ -199,7 +199,7 @@ func NewFramework(inClusterTestSuite bool) *Framework {
                }

                var err error
-               contourBin, err = gexec.Build("github.com/projectcontour/contour/cmd/contour")
+               contourBin, err = gexec.Build("github.com/projectcontour/contour/cmd/contour", "-race")
                require.NoError(t, err)
        }

```

then run the e2e tests with something like `CONTOUR_E2E_PACKAGE_FOCUS=./test/e2e/httpproxy make run-e2e`

the tests won't fail but you will see the data race in the contour logs that are printed

no need to commit the diffs I've provided here, I'm working on adding to the e2e test framework to detect data races automatically

Reply:
It should be okay now.